### PR TITLE
Closes #36541 (BUG: ValueError: cannot convert float NaN to integer when resetting MultiIndex with NaT values)

### LINKED ISSUE #36541: BUG: ValueError: cannot convert float NaN to integer when resetting MultiIndex with NaT values
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [x] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
In [18]: ix = pd.MultiIndex.from_tuples([(pd.NaT, 1), (pd.NaT, 2)], names=['a', 'b'])

In [19]: ix
Out[19]: 
MultiIndex([('NaT', 1),
            ('NaT', 2)],
           names=['a', 'b'])

In [20]: d = pd.DataFrame({'x': [11, 12]}, index=ix)

In [21]: d
Out[21]: 
        x
a   b    
NaT 1  11
    2  12

In [22]: d.reset_index()
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-22-4653618060e8> in <module>
----> 1 d.reset_index()

~/envs/pandas-test/lib/python3.8/site-packages/pandas/core/frame.py in reset_index(self, level, drop, inplace, col_level, col_fill)
   4851                     name = tuple(name_lst)
   4852                 # to ndarray and maybe infer different dtype
-> 4853                 level_values = _maybe_casted_values(lev, lab)
   4854                 new_obj.insert(0, name, level_values)
   4855 

~/envs/pandas-test/lib/python3.8/site-packages/pandas/core/frame.py in _maybe_casted_values(index, labels)
   4784                     dtype = index.dtype
   4785                     fill_value = na_value_for_dtype(dtype)
-> 4786                     values = construct_1d_arraylike_from_scalar(
   4787                         fill_value, len(mask), dtype
   4788                     )

~/envs/pandas-test/lib/python3.8/site-packages/pandas/core/dtypes/cast.py in construct_1d_arraylike_from_scalar(value, length, dtype)
   1556 
   1557         subarr = np.empty(length, dtype=dtype)
-> 1558         subarr.fill(value)
   1559 
   1560     return subarr

ValueError: cannot convert float NaN to integer
```

#### Problem description

With the introduction and use of `groupby(..., dropna=False)` multiindex with NaT values are more likely to occur which exhibits a few issues that previously went undetected. This issue was discovered when finding a workaround for another `dropna=False` related issue (https://github.com/pandas-dev/pandas/issues/36060#issuecomment-695869490)

Further investigation shows that this may be an issue with numpy not accepting pd.NaT. The following code reproduces the issue in `construct_1d_arraylike_from_scalar`:

```
In [33]: a = np.empty(2, dtype='datetime64[ns]')

In [34]: a.fill(pd.NaT)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-34-9fa9ff66da99> in <module>
----> 1 a.fill(pd.NaT)

ValueError: cannot convert float NaN to integer
```

which led me to propose this fix:

```diff
--- a/pandas/core/dtypes/cast.py
+++ b/pandas/core/dtypes/cast.py
@@ -1559,6 +1559,12 @@ def construct_1d_arraylike_from_scalar(
             dtype = np.dtype("object")
             if not isna(value):
                 value = ensure_str(value)
+        elif isinstance(dtype, np.dtype) and dtype.kind == "M" and value is NaT:
+            # can't fill sub array directly with pandas' NaT:
+            #
+            # > a.fill(pd.NaT)
+            # ValueError: cannot convert float NaN to integer
+            value = np.datetime64("NaT")
 
         subarr = np.empty(length, dtype=dtype)
         subarr.fill(value)
```

the `value is NaT` check could possibly be extended to `isna(value)`...

#### Expected Output

No ValueError being raised

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : 15539fa6217039c2e5cb70b1d942caec9c5a2de3
python           : 3.8.5.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 19.6.0
Version          : Darwin Kernel Version 19.6.0: Thu Jun 18 20:49:00 PDT 2020; root:xnu-6153.141.1~1/RELEASE_X86_64
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : en_AU.UTF-8

pandas           : 1.2.0.dev0+453.g15539fa62.dirty
numpy            : 1.19.2
pytz             : 2020.1
dateutil         : 2.8.1
pip              : 20.1.1
setuptools       : 46.4.0
Cython           : 0.29.21
pytest           : 6.0.2
hypothesis       : 5.35.3
sphinx           : 3.2.1
blosc            : 1.9.2
feather          : None
xlsxwriter       : 1.3.4
lxml.etree       : 4.5.2
html5lib         : 1.1
pymysql          : None
psycopg2         : None
jinja2           : 2.11.2
IPython          : 7.18.1
pandas_datareader: None
bs4              : 4.9.1
bottleneck       : 1.3.2
fsspec           : 0.8.2
fastparquet      : 0.4.1
gcsfs            : 0.7.1
matplotlib       : 3.3.2
numexpr          : 2.7.1
odfpy            : None
openpyxl         : 3.0.5
pandas_gbq       : None
pyarrow          : 1.0.1
pytables         : None
pyxlsb           : None
s3fs             : 0.5.1
scipy            : 1.5.2
sqlalchemy       : 1.3.19
tables           : 3.6.1
tabulate         : 0.8.7
xarray           : 0.16.1
xlrd             : 1.2.0
xlwt             : 1.3.0
numba            : 0.51.2

</details>

### PR DESCRIPTION
**Pull request by ssche**
- [x] closes #36541
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v1.2.0.rst [hunk 1]
```diff
@@ -310,6 +310,7 @@ MultiIndex
 ^^^^^^^^^^
 
 - Bug in :meth:`DataFrame.xs` when used with :class:`IndexSlice` raises ``TypeError`` with message `Expected label or tuple of labels` (:issue:`35301`)
+- Bug in :func:`construct_1d_arraylike_from_scalar` when using :meth:`DataFrame.reset_index` with ``NaT`` values in index raises ``ValueError`` with message `cannot convert float NaN to integer` (:issue:`36541`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.2.0.rst:313]: remove the first part - these are always user facing notes

File: pandas/core/dtypes/cast.py [hunk 2]
```diff
@@ -1559,6 +1559,11 @@ def construct_1d_arraylike_from_scalar(
             dtype = np.dtype("object")
             if not isna(value):
                 value = ensure_str(value)
+        elif isinstance(dtype, np.dtype) and dtype.kind == "M" and value is NaT:
```
Comment (jreback) [inline pandas/core/dtypes/cast.py:1562]: use is_nat_for_dtype (is not exactly this same but similar)

File: pandas/tests/series/indexing/test_multiindex.py [hunk 3]
```diff
@@ -20,3 +20,13 @@ def test_access_none_value_in_multiindex():
...
+
+
+def test_nat_multi_index():
+    # GH36541: that reset_index() does not raise ValueError
+    ix = pd.MultiIndex.from_tuples([(pd.NaT, 1), (pd.NaT, 2)], names=["a", "b"])
+    result = pd.DataFrame({"x": [11, 12]}, index=ix)
+    result = result.reset_index()
+
+    expected = pd.DataFrame({"a": [pd.NaT, pd.NaT], "b": [1, 2], "x": [11, 12]})
+    pd.testing.assert_frame_equal(result, expected)
```
Comment (jreback) [inline pandas/tests/series/indexing/test_multiindex.py:32]: use tm.assert...

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/dtypes/cast.py (hunk 2 above)
Comment (jreback) [reply pandas/core/dtypes/cast.py]: is_valid_nat_for_dtype is the function

Comment (ssche) [reply pandas/core/dtypes/cast.py]: I couldn't just use `is_valid_nat_for_dtype` as it also returned true for `is_valid_nat_for_dtype(nan, float64)` (which was a problem for the `test_modulo2` test case for instance (I didn't check others which also failed)).

**Comment by ssche** [comment]
I might need some help regarding the build error (Arm64, Xenial, Py3.7). Seems lime there are some files missing, but I’m not sure what I could do to fix that...?

File: pandas/tests/series/indexing/test_multiindex.py [hunk 4]
```diff
@@ -20,3 +21,13 @@ def test_access_none_value_in_multiindex():
     s = Series([1] * len(midx), dtype=object, index=midx)
     result = s.loc[("Level1", "Level2_a")]
     assert result == 1
+
+
+def test_nat_multi_index():
+    # GH36541: that reset_index() does not raise ValueError
+    ix = pd.MultiIndex.from_tuples([(pd.NaT, 1), (pd.NaT, 2)], names=["a", "b"])
```
Comment (jreback) [inline pandas/tests/series/indexing/test_multiindex.py:28]: can you try this with one element a Timestamp (and one a NaT)
also add a case where you have a Timedelta and a NaT (and I think we might need to change the code slightly)

can parameterize the test.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/series/indexing/test_multiindex.py (hunk 4 above)
Comment (ssche) [reply pandas/tests/series/indexing/test_multiindex.py]: added test cases. please have another look and advise.

**Review by jreback — APPROVED** [review]

File: pandas/core/dtypes/cast.py [hunk 5]
```diff
@@ -1559,6 +1559,11 @@ def construct_1d_arraylike_from_scalar(
             dtype = np.dtype("object")
             if not isna(value):
                 value = ensure_str(value)
+        elif dtype.kind == "M" and is_valid_nat_for_dtype(value, dtype):
```
Comment (jreback) [inline pandas/core/dtypes/cast.py:1562]: should be ```dtype.kind in ['M', 'm']``` (do the tests fail w/o this?)

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
pls merge master as well

File: pandas/core/dtypes/cast.py (hunk 5 above)
Comment (ssche) [reply pandas/core/dtypes/cast.py]: no, they didn't before with "M" only (except for the ARM tests which I believe have an unrelated problem).

Trying with timedelta "m" now...

**Comment by ssche** [comment]
ready to merge @jreback

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @ssche very nice!

**Pull request merged**